File: ra-intelligence/discovery/benchmark/evaluator.py
from __future__ import annotations

import re
from typing import Any

from discovery.benchmark.models import DiscoveryBenchmarkScenario
from discovery.benchmark.results import AutomaticScore
from discovery.models import DiscoveryProviderResult, DiscoveryState


STOPWORDS = {
    "the",
    "and",
    "for",
    "with",
    "that",
    "this",
    "their",
    "what",
    "from",
    "into",
    "immediately",
    "again",
    "asking",
    "full",
    "one",
}


def normalize(value: str) -> str:
    value = value.lower()
    value = re.sub(r"[^a-z0-9\s]", " ", value)
    value = re.sub(r"\s+", " ", value)
    return value.strip()


def field_is_captured(
    state: DiscoveryState,
    field_name: str,
) -> bool:
    value = getattr(state, field_name)

    if isinstance(value, list):
        return len(value) > 0

    if isinstance(value, str):
        return bool(value.strip())

    if isinstance(value, (int, float)):
        return True

    return value is not None


def topic_tokens(topic: str) -> list[str]:
    return [
        token
        for token in normalize(topic).split()
        if len(token) >= 3 and token not in STOPWORDS
    ]


def mentions_topic(
    text: str,
    topic: str,
) -> bool:
    normalized_text = normalize(text)
    tokens = topic_tokens(topic)

    if not tokens:
        return False

    hits = sum(
        1
        for token in tokens
        if token in normalized_text
    )

    required_hits = max(
        1,
        (len(tokens) + 1) // 2,
    )

    return hits >= required_hits


def values_preserved(
    expected: Any,
    actual: Any,
) -> bool:
    if isinstance(expected, list):
        if not isinstance(actual, list):
            return False

        normalized_actual = [
            normalize(item)
            if isinstance(item, str)
            else item
            for item in actual
        ]

        for item in expected:
            normalized_expected = (
                normalize(item)
                if isinstance(item, str)
                else item
            )

            if normalized_expected not in normalized_actual:
                return False

        return True

    if isinstance(expected, str) and isinstance(actual, str):
        expected_text = normalize(expected)
        actual_text = normalize(actual)

        return (
            expected_text == actual_text
            or expected_text in actual_text
            or actual_text in expected_text
        )

    return expected == actual


def score_business_understanding(
    scenario: DiscoveryBenchmarkScenario,
    result: DiscoveryProviderResult,
    notes: list[str],
) -> float:
    score = 0.0

    accepted_services = (
        scenario.expected.acceptable_service_categories
        or [scenario.expected.service_category]
    )

    if (
        result.response.state.service_category
        in accepted_services
    ):
        score += 10
    else:
        notes.append(
            "Service mismatch: "
            f"expected one of {accepted_services}, "
            f"received {result.response.state.service_category}."
        )

    required_fields = [
        field_name
        for field_name in scenario.expected.must_capture
        if field_name != "service_category"
    ]

    if not required_fields:
        return score + 10

    captured = sum(
        1
        for field_name in required_fields
        if field_is_captured(
            result.response.state,
            field_name,
        )
    )

    score += (captured / len(required_fields)) * 10

    if captured < len(required_fields):
        notes.append(
            f"Captured {captured}/{len(required_fields)} "
            "expected discovery fields."
        )

    return score


def score_discovery_efficiency(
    scenario: DiscoveryBenchmarkScenario,
    result: DiscoveryProviderResult,
    notes: list[str],
) -> float:
    score = 0.0
    reply = result.response.reply

    question_count = reply.count("?")

    if scenario.expected.complete:
        if question_count == 0:
            score += 10
        else:
            notes.append(
                "Expected discovery to stop, but response "
                f"asked {question_count} question(s)."
            )

    elif question_count <= 1:
        score += 10

    elif question_count == 2:
        score += 5
        notes.append(
            "Response asked two questions; Ra preference "
            "is one primary question per turn."
        )

    else:
        notes.append(
            f"Response asked {question_count} questions in one turn."
        )

    violations = [
        topic
        for topic in scenario.expected.avoid_question_topics
        if mentions_topic(reply, topic)
    ]

    if not violations:
        score += 10

    elif len(violations) == 1:
        score += 5
        notes.append(
            f"Potential unnecessary topic: {violations[0]}."
        )

    else:
        notes.append(
            "Multiple unnecessary topics detected: "
            + ", ".join(violations)
            + "."
        )

    return score


def score_state_accuracy(
    scenario: DiscoveryBenchmarkScenario,
    result: DiscoveryProviderResult,
    notes: list[str],
) -> float:
    score = 0.0

    if result.response.complete == scenario.expected.complete:
        score += 5
    else:
        notes.append(
            "Completion mismatch: "
            f"expected {scenario.expected.complete}, "
            f"received {result.response.complete}."
        )

    expected_next = scenario.expected.recommended_next_step

    if expected_next is None:
        score += 5

    elif result.response.recommended_next_step == expected_next:
        score += 5

    else:
        notes.append(
            "Next-step mismatch: "
            f"expected {expected_next}, "
            f"received {result.response.recommended_next_step}."
        )

    known_entries = list(
        scenario.current_state.items()
    )

    if not known_entries:
        score += 5
        return score

    output_state = result.response.state.model_dump()

    preserved = sum(
        1
        for key, value in known_entries
        if values_preserved(
            value,
            output_state.get(key),
        )
    )

    score += (
        preserved / len(known_entries)
    ) * 5

    if preserved < len(known_entries):
        notes.append(
            f"Preserved {preserved}/{len(known_entries)} "
            "previously established state fields."
        )

    return score


def score_latency(
    latency_ms: float,
) -> float:
    if latency_ms <= 2500:
        return 3

    if latency_ms <= 5000:
        return 2

    if latency_ms <= 8000:
        return 1

    return 0


def score_cost(
    estimated_cost_usd: float | None,
) -> float:
    if estimated_cost_usd is None:
        return 0

    if estimated_cost_usd <= 0.01:
        return 2

    if estimated_cost_usd <= 0.03:
        return 1.5

    if estimated_cost_usd <= 0.10:
        return 1

    return 0.5


def evaluate_automatically(
    scenario: DiscoveryBenchmarkScenario,
    result: DiscoveryProviderResult,
) -> AutomaticScore:
    notes: list[str] = []

    business_understanding = score_business_understanding(
        scenario,
        result,
        notes,
    )

    discovery_efficiency = score_discovery_efficiency(
        scenario,
        result,
        notes,
    )

    state_accuracy = score_state_accuracy(
        scenario,
        result,
        notes,
    )

    latency = score_latency(
        result.latency_ms,
    )

    cost = score_cost(
        result.estimated_cost_usd,
    )

    total = (
        business_understanding
        + discovery_efficiency
        + state_accuracy
        + latency
        + cost
    )

    return AutomaticScore(
        business_understanding=business_understanding,
        discovery_efficiency=discovery_efficiency,
        state_accuracy=state_accuracy,
        latency=latency,
        cost=cost,
        total=total,
        notes=notes,
    )

File: src/lib/discovery/benchmark/evaluator.ts
import type {
  DiscoveryState,
} from "../types";

import type {
  DiscoveryProviderResult,
} from "../providers/provider";

import type {
  DiscoveryBenchmarkScenario,
} from "./types";

import type {
  AutomaticScore,
} from "./results";

const STOPWORDS = new Set([
  "the",
  "and",
  "for",
  "with",
  "that",
  "this",
  "their",
  "what",
  "from",
  "into",
  "immediately",
  "again",
  "asking",
  "full",
  "one",
]);

function normalize(value: string): string {
  return value
    .toLowerCase()
    .replace(/[^a-z0-9\s]/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function fieldIsCaptured(
  state: DiscoveryState,
  field: keyof DiscoveryState,
): boolean {
  const value = state[field];

  if (Array.isArray(value)) {
    return value.length > 0;
  }

  if (typeof value === "string") {
    return value.trim().length > 0;
  }

  if (typeof value === "number") {
    return Number.isFinite(value);
  }

  return value !== null && value !== undefined;
}

function topicTokens(topic: string): string[] {
  return normalize(topic)
    .split(" ")
    .filter(
      (token) =>
        token.length >= 3 &&
        !STOPWORDS.has(token),
    );
}

function mentionsTopic(
  text: string,
  topic: string,
): boolean {
  const normalizedText = normalize(text);
  const tokens = topicTokens(topic);

  if (tokens.length === 0) {
    return false;
  }

  const hits = tokens.filter((token) =>
    normalizedText.includes(token),
  ).length;

  return hits >= Math.max(
    1,
    Math.ceil(tokens.length * 0.5),
  );
}

function scoreBusinessUnderstanding(
  scenario: DiscoveryBenchmarkScenario,
  result: DiscoveryProviderResult,
  notes: string[],
): number {
  let score = 0;

  if (
    result.response.state.serviceCategory ===
    scenario.expected.serviceCategory
  ) {
    score += 10;
  } else {
    notes.push(
      `Service mismatch: expected ${scenario.expected.serviceCategory}, received ${result.response.state.serviceCategory}.`,
    );
  }

  const requiredFields =
    scenario.expected.mustCapture.filter(
      (field) => field !== "serviceCategory",
    );

  if (requiredFields.length === 0) {
    return score + 10;
  }

  const captured = requiredFields.filter(
    (field) =>
      fieldIsCaptured(
        result.response.state,
        field,
      ),
  ).length;

  score +=
    (captured / requiredFields.length) * 10;

  if (captured < requiredFields.length) {
    notes.push(
      `Captured ${captured}/${requiredFields.length} expected discovery fields.`,
    );
  }

  return score;
}

function scoreDiscoveryEfficiency(
  scenario: DiscoveryBenchmarkScenario,
  result: DiscoveryProviderResult,
  notes: string[],
): number {
  let score = 0;

  const reply = result.response.reply;
  const questionCount =
    (reply.match(/\?/g) ?? []).length;

  if (scenario.expected.complete) {
    if (questionCount === 0) {
      score += 10;
    } else {
      notes.push(
        `Expected discovery to stop, but response asked ${questionCount} question(s).`,
      );
    }
  } else if (questionCount <= 1) {
    score += 10;
  } else if (questionCount === 2) {
    score += 5;
    notes.push(
      "Response asked two questions; Paradigm preference is one primary question per turn.",
    );
  } else {
    notes.push(
      `Response asked ${questionCount} questions in one turn.`,
    );
  }

  const violations =
    scenario.expected.avoidQuestionTopics.filter(
      (topic) =>
        mentionsTopic(reply, topic),
    );

  if (violations.length === 0) {
    score += 10;
  } else if (violations.length === 1) {
    score += 5;
    notes.push(
      `Potential unnecessary topic: ${violations[0]}.`,
    );
  } else {
    notes.push(
      `Multiple unnecessary topics detected: ${violations.join(", ")}.`,
    );
  }

  return score;
}

function valuesPreserved(
  expected: unknown,
  actual: unknown,
): boolean {
  if (Array.isArray(expected)) {
    if (!Array.isArray(actual)) {
      return false;
    }

    const normalizedActual =
      actual.map((item) =>
        typeof item === "string"
          ? normalize(item)
          : item,
      );

    return expected.every((item) => {
      const normalizedExpected =
        typeof item === "string"
          ? normalize(item)
          : item;

      return normalizedActual.includes(
        normalizedExpected,
      );
    });
  }

  if (
    typeof expected === "string" &&
    typeof actual === "string"
  ) {
    const a = normalize(expected);
    const b = normalize(actual);

    return a === b || a.includes(b) || b.includes(a);
  }

  return expected === actual;
}

function scoreStateAccuracy(
  scenario: DiscoveryBenchmarkScenario,
  result: DiscoveryProviderResult,
  notes: string[],
): number {
  let score = 0;

  if (
    result.response.complete ===
    scenario.expected.complete
  ) {
    score += 5;
  } else {
    notes.push(
      `Completion mismatch: expected ${scenario.expected.complete}, received ${result.response.complete}.`,
    );
  }

  if (
    scenario.expected.recommendedNextStep
  ) {
    if (
      result.response.recommendedNextStep ===
      scenario.expected.recommendedNextStep
    ) {
      score += 5;
    } else {
      notes.push(
        `Next-step mismatch: expected ${scenario.expected.recommendedNextStep}, received ${result.response.recommendedNextStep}.`,
      );
    }
  } else {
    score += 5;
  }

  const knownEntries =
    Object.entries(scenario.currentState);

  if (knownEntries.length === 0) {
    score += 5;
    return score;
  }

  const outputState =
    result.response.state as unknown as Record<
      string,
      unknown
    >;

  const preserved = knownEntries.filter(
    ([key, value]) =>
      valuesPreserved(
        value,
        outputState[key],
      ),
  ).length;

  score +=
    (preserved / knownEntries.length) * 5;

  if (preserved < knownEntries.length) {
    notes.push(
      `Preserved ${preserved}/${knownEntries.length} previously established state fields.`,
    );
  }

  return score;
}

function scoreLatency(
  latencyMs: number,
): number {
  if (latencyMs <= 2500) return 3;
  if (latencyMs <= 5000) return 2;
  if (latencyMs <= 8000) return 1;

  return 0;
}

function scoreCost(
  estimatedCostUsd?: number,
): number {
  if (
    estimatedCostUsd === undefined ||
    estimatedCostUsd === null
  ) {
    return 0;
  }

  if (estimatedCostUsd <= 0.01) return 2;
  if (estimatedCostUsd <= 0.03) return 1.5;
  if (estimatedCostUsd <= 0.10) return 1;

  return 0.5;
}

export function evaluateAutomatically(
  scenario: DiscoveryBenchmarkScenario,
  result: DiscoveryProviderResult,
): AutomaticScore {
  const notes: string[] = [];

  const businessUnderstanding =
    scoreBusinessUnderstanding(
      scenario,
      result,
      notes,
    );

  const discoveryEfficiency =
    scoreDiscoveryEfficiency(
      scenario,
      result,
      notes,
    );

  const stateAccuracy =
    scoreStateAccuracy(
      scenario,
      result,
      notes,
    );

  const latency = scoreLatency(
    result.latencyMs,
  );

  const cost = scoreCost(
    result.estimatedCostUsd,
  );

  const total =
    businessUnderstanding +
    discoveryEfficiency +
    stateAccuracy +
    latency +
    cost;

  return {
    businessUnderstanding,
    discoveryEfficiency,
    stateAccuracy,
    latency,
    cost,
    total,
    maxScore: 60,
    notes,
  };
}
